framework/src/main/java/org/checkerframework/framework/type/GenericAnnotatedTypeFactory.java

```diff
@@ -162,6 +162,29 @@
      */
     private boolean shouldDefaultTypeVarLocals;
 
+    /**
+     * Elements representing variables for which the type of the initializer is being determined in
+     * order to apply qualifier parameter defaults.
+     *
+     * <p>Local variables with a qualifier parameter get their declared type from the type of their
+     * initializer. Sometimes the initializer's type depends on the type of the variable, such as
+     * during type variable inference or when a variable is used in its own initializer as in
+     * "Object o = (o = null)". This creates a circular dependency resulting in infinite recursion.
+     * To prevent this, variables in this set should not be typed based on their initializer, but by
+     * using normal defaults.
+     *
+     * @see GenericAnnotatedTypeFactory#applyLocalVariableQualifierParameterDefaults
+     */
+    private Set<Element> variablesUnderInitialization;
+
+    /**
+     * Caches types of initializers for local variables with a qualifier parameter, so that they
+     * aren't computed each time the type of a variable is looked up.
+     *
+     * @see GenericAnnotatedTypeFactory#applyLocalVariableQualifierParameterDefaults
+     */
+    private Map<Tree, AnnotatedTypeMirror> initializerCache;
```

Comment:
Did you have to add this cache so that the Daikon test would not time-out?  We have avoided caching expressions because dataflow can change the type of an expression; see https://github.com/typetools/checker-framework/issues/602.  I don't think this cache makes the problem worse than it already is though.

Reply:
Yes, the Daikon test timed out before Jason added this cache.

Reply:
I actually added two optimizations. The first was to not calculate the type of the initializer if the type of the local variable it's assigned to has no qualifier parameter in any hierarchy. Previously, it calculated the type of the initializer for every local variable and then filtered to retain only the annotations in hierarchies with a qualifier parameter. The second  optimization was to add a cache.

The first fix alone would have been sufficient to fix the CI because it's likely that no types in the daikon test use a qualifier parameter. However, it indicates that there's some performance penalty for this check when used. Retaining the cache would probably be helpful for situations when writing `@HasQualifierParameter` on a package, which applies it to all classes in the package.

Reply:
Thanks for the explanation!